api/label/encoder.go

```diff
@@ -0,0 +1,147 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package label
+
+import (
+	"bytes"
+	"sync"
+	"sync/atomic"
+
+	"go.opentelemetry.io/otel/api/core"
+)
+
+type (
+	// Encoder is a mechanism.
+	Encoder interface {
+		// Encode is called (concurrently) in instrumentation context.
+		//
+		// The expectation is that when setting up an export pipeline
+		// both the batcher and the exporter will use the same label
+		// encoder to avoid the duplicate computation of the encoded
+		// labels in the export path.
+		Encode(Iterator) string
+
+		// ID returns a value that is unique for each class of
+		// label encoder.  Label encoders allocate these using
+		// `NewEncoderID`.
+		ID() EncoderID
+	}
+
+	// EncoderID
+	EncoderID struct {
+		value int64
+	}
+
+	defaultLabelEncoder struct {
+		// pool is a pool of labelset builders.  The buffers in this
+		// pool grow to a size that most label encodings will not
+		// allocate new memory.
+		pool sync.Pool // *bytes.Buffer
+	}
+)
+
+// escapeChar is used to ensure uniqueness of the label encoding where
+// keys or values contain either '=' or ','.  Since there is no parser
+// needed for this encoding and its only requirement is to be unique,
+// this choice is arbitrary.  Users will see these in some exporters
+// (e.g., stdout), so the backslash ('\') is used a conventional choice.
+const escapeChar = '\\'
+
+var (
+	_ Encoder = &defaultLabelEncoder{}
+
+	// labelEncoderIDCounter is for generating IDs for other label
+	// encoders.
+	encoderIDCounter int64 = 1
```

Comment:
The whole point is that zero is invalid. But counter can start at zero, because the first call to Atomic.AddInt64 will return 1, not zero.

Reply:
> The whole point is that zero is invalid. But counter can start at zero, because the first call to Atomic.AddInt64 will return 1, not zero.

Gotcha :+1: